File: rust/var-core/src/scanner.rs
//! Turns raw Markdown into a flat list of [`Block`] nodes — port of `scanner.ts`
//! / `Scanner.java`. Offsets in stored spans are UTF-16 code units; the line
//! splitter keeps a running (byte, UTF-16) dual cursor and per-line regex offsets
//! are converted from bytes to UTF-16.

use crate::ast::{
    Block, Blockquote, Fence, Heading, ListItem, Paragraph, Row, SegmentOffset, Table,
    ThematicBreak,
};
use crate::offsets::{java_trim, utf16_index, utf16_len};
use crate::span::Span;
use crate::table_cells::parse_row_cells;
use regex::Regex;
use std::sync::LazyLock;

/// One line of source, with its UTF-16 and byte offsets in the full source.
struct RawLine {
    text: String,
    start_offset: usize,
    end_offset: usize,
    start_byte: usize,
    end_byte: usize,
}

// `\1` backreference is expanded into three alternatives (the `regex` crate has
// no backreferences); otherwise these mirror the Java patterns. `[0-9]` keeps the
// ordered-list digit class ASCII.
static THEMATIC_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^\s*(?:-(?:\s*-){2,}|\*(?:\s*\*){2,}|_(?:\s*_){2,})\s*$").unwrap()
});
static UL_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^(\s*)([-*+])\s+(.*)$").unwrap());
static OL_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^(\s*)([0-9]+)([.)])\s+(.*)$").unwrap());
static BQ_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^>\s?(.*)$").unwrap());
static FENCE_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^(`{3,})\s*(\S*)\s*$").unwrap());
static ROW_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^\|(.+)\|\s*$").unwrap());
static DELIM_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^\|\s*:?-+:?\s*(\|\s*:?-+:?\s*)*\|\s*$").unwrap());
static HEADING_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^(#{1,6})\s+(.*?)(?:\s+#+)?\s*$").unwrap());
static HEADING_PREFIX_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^#{1,6}\s+").unwrap());

/// Scans `source` into a list of [`Block`] nodes.
pub fn scan(source: &str) -> Vec<Block> {
    let lines = split_lines(source);
    let mut blocks = Vec::new();

    let mut i = 0;
    while i < lines.len() {
        if java_trim(&lines[i].text).is_empty() {
            i += 1;
            continue;
        }
        if let Some((fence, next)) = try_fence(source, &lines, i) {
            blocks.push(Block::Fence(fence));
            i = next;
            continue;
        }
        if let Some((table, next)) = try_table(source, &lines, i) {
            blocks.push(Block::Table(table));
            i = next;
            continue;
        }
        if let Some(tb) = try_thematic_break(source, &lines[i]) {
            blocks.push(Block::ThematicBreak(tb));
            i += 1;
            continue;
        }
        if let Some((quote, next)) = try_blockquote(source, &lines, i) {
            blocks.push(Block::Blockquote(quote));
            i = next;
            continue;
        }
        if let Some(heading) = try_heading(source, &lines[i]) {
            blocks.push(Block::Heading(heading));
            i += 1;
            continue;
        }
        if let Some(item) = try_list_item(source, &lines[i]) {
            blocks.push(Block::ListItem(item));
            i += 1;
            continue;
        }
        let (paragraph, next) = consume_paragraph(source, &lines, i);
        blocks.push(Block::Paragraph(paragraph));
        i = next;
    }
    blocks
}

fn split_lines(source: &str) -> Vec<RawLine> {
    let mut out = Vec::new();
    let mut byte_start = 0;
    let mut u16_start = 0;
    let mut u16 = 0;
    for (byte_i, c) in source.char_indices() {
        if c == '\n' {
            out.push(RawLine {
                text: source[byte_start..byte_i].to_string(),
                start_offset: u16_start,
                end_offset: u16,
                start_byte: byte_start,
                end_byte: byte_i,
            });
            byte_start = byte_i + 1;
            u16_start = u16 + 1;
        }
        u16 += c.len_utf16();
    }
    out.push(RawLine {
        text: source[byte_start..].to_string(),
        start_offset: u16_start,
        end_offset: u16,
        start_byte: byte_start,
        end_byte: source.len(),
    });
    out
}

fn try_thematic_break(source: &str, line: &RawLine) -> Option<ThematicBreak> {
    if !THEMATIC_RE.is_match(&line.text) {
        return None;
    }
    Some(ThematicBreak {
        span: Span::from_offsets(source, line.start_offset, line.end_offset),
    })
}

fn try_heading(source: &str, line: &RawLine) -> Option<Heading> {
    let m = HEADING_RE.captures(&line.text)?;
    let hashes = m.get(1).unwrap().as_str();
    let text = java_trim(m.get(2).unwrap().as_str()).to_string();
    Some(Heading {
        level: hashes.len(),
        text,
        span: Span::from_offsets(source, line.start_offset, line.end_offset),
    })
}

fn try_list_item(source: &str, line: &RawLine) -> Option<ListItem> {
    if let Some(ul) = UL_RE.captures(&line.text) {
        let text = ul.get(3).unwrap().as_str();
        let marker_start = line.start_offset + utf16_len(ul.get(1).unwrap().as_str());
        let marker_end = marker_start + utf16_len(ul.get(2).unwrap().as_str());
        let text_start = line.start_offset + utf16_index(&line.text, line.text.find(text).unwrap());
        return Some(ListItem {
            text: text.to_string(),
            span: Span::from_offsets(source, line.start_offset, line.end_offset),
            segment_map: vec![SegmentOffset::new(0, text_start)],
            ordered: false,
            marker_span: Span::from_offsets(source, marker_start, marker_end),
        });
    }
    if let Some(ol) = OL_RE.captures(&line.text) {
        let text = ol.get(4).unwrap().as_str();
        let marker_start = line.start_offset + utf16_len(ol.get(1).unwrap().as_str());
        let marker_end = marker_start
            + utf16_len(ol.get(2).unwrap().as_str())
            + utf16_len(ol.get(3).unwrap().as_str());
        let text_start = line.start_offset + utf16_index(&line.text, line.text.find(text).unwrap());
        return Some(ListItem {
            text: text.to_string(),
            span: Span::from_offsets(source, line.start_offset, line.end_offset),
            segment_map: vec![SegmentOffset::new(0, text_start)],
            ordered: true,
            marker_span: Span::from_offsets(source, marker_start, marker_end),
        });
    }
    None
}

fn try_blockquote(
    source: &str,
    lines: &[RawLine],
    start_idx: usize,
) -> Option<(Blockquote, usize)> {
    let first = &lines[start_idx];
    let m = BQ_RE.captures(&first.text)?;
    let first_segment = m.get(1).unwrap().as_str().to_string();

    let mut segments = vec![first_segment.clone()];
    let mut segment_map = vec![SegmentOffset::new(
        0,
        first.start_offset + utf16_index(&first.text, first.text.find(&first_segment).unwrap()),
    )];
    let mut joined_text_offset = utf16_len(&first_segment);

    let mut i = start_idx + 1;
    let mut end_offset = first.end_offset;
    while i < lines.len() {
        let ln = &lines[i];
        let Some(next) = BQ_RE.captures(&ln.text) else {
            break;
        };
        let segment = next.get(1).unwrap().as_str().to_string();
        joined_text_offset += 1; // newline separator
        segment_map.push(SegmentOffset::new(
            joined_text_offset,
            ln.start_offset + utf16_index(&ln.text, ln.text.find(&segment).unwrap()),
        ));
        joined_text_offset += utf16_len(&segment);
        segments.push(segment);
        end_offset = ln.end_offset;
        i += 1;
    }
    let quote = Blockquote {
        text: segments.join("\n"),
        span: Span::from_offsets(source, first.start_offset, end_offset),
        segment_map,
    };
    Some((quote, i))
}

fn consume_paragraph(source: &str, lines: &[RawLine], start_idx: usize) -> (Paragraph, usize) {
    let first = &lines[start_idx];
    let mut end_idx = start_idx;
    while end_idx + 1 < lines.len() {
        let candidate = &lines[end_idx + 1];
        let t = &candidate.text;
        if java_trim(t).is_empty()
            || HEADING_PREFIX_RE.is_match(t)
            || UL_RE.is_match(t)
            || OL_RE.is_match(t)
            || BQ_RE.is_match(t)
            || FENCE_RE.is_match(t)
            || ROW_RE.is_match(t)
            || THEMATIC_RE.is_match(t)
        {
            break;
        }
        end_idx += 1;
    }
    let last = &lines[end_idx];
    let paragraph = Paragraph {
        text: source[first.start_byte..last.end_byte].to_string(),
        span: Span::from_offsets(source, first.start_offset, last.end_offset),
        segment_map: vec![SegmentOffset::new(0, first.start_offset)],
    };
    (paragraph, end_idx + 1)
}

fn try_fence(source: &str, lines: &[RawLine], start_idx: usize) -> Option<(Fence, usize)> {
    let start = &lines[start_idx];
    let open = FENCE_RE.captures(&start.text)?;
    let fence_marker = open.get(1).unwrap().as_str().to_string();
    let info = java_trim(open.get(2).unwrap().as_str()).to_string();

    let mut i = start_idx + 1;
    let mut body_start: Option<(usize, usize)> = None; // (u16, byte)
    let mut body_end: Option<(usize, usize)> = None;
    let mut end_offset = start.end_offset;
    while i < lines.len() {
        let ln = &lines[i];
        if let Some(close) = FENCE_RE.captures(&ln.text) {
            if close.get(1).unwrap().as_str().len() >= fence_marker.len() {
                end_offset = ln.end_offset;
                break;
            }
        }
        if body_start.is_none() {
            body_start = Some((ln.start_offset, ln.start_byte));
        }
        // Include the newline that separates this line from the next.
        body_end = Some((ln.end_offset + 1, ln.end_byte + 1));
        i += 1;
    }

    let source_u16 = utf16_len(source);
    let clamped_end_u16 = body_end.map_or(0, |(u16, _)| u16.min(source_u16));
    let clamped_end_byte = body_end.map_or(0, |(_, byte)| byte.min(source.len()));
    let body = match (body_start, body_end) {
        (Some((_, sb)), Some(_)) => source[sb..clamped_end_byte].to_string(),
        _ => String::new(),
    };
    let fallback = start.end_offset;
    let body_span = Span::from_offsets(
        source,
        body_start.map_or(fallback, |(u16, _)| u16),
        if body_end.is_some() {
            clamped_end_u16
        } else {
            fallback
        },
    );
    let fence = Fence {
        span: Span::from_offsets(source, start.start_offset, end_offset),
        info,
        body,
        body_span,
    };
    Some((fence, i + 1))
}

fn try_table(source: &str, lines: &[RawLine], start_idx: usize) -> Option<(Table, usize)> {
    if start_idx + 1 >= lines.len() {
        return None;
    }
    let header_line = &lines[start_idx];
    let delim_line = &lines[start_idx + 1];
    if !ROW_RE.is_match(&header_line.text) || !DELIM_RE.is_match(&delim_line.text) {
        return None;
    }

    let header_parsed = parse_row_cells(&header_line.text, header_line.start_offset, source);
    let header = Row {
        cells: header_parsed.cells,
        cell_spans: header_parsed.cell_spans,
        span: Span::from_offsets(source, header_line.start_offset, header_line.end_offset),
    };

    let mut rows = Vec::new();
    let mut i = start_idx + 2;
    while i < lines.len() {
        let ln = &lines[i];
        if !ROW_RE.is_match(&ln.text) {
            break;
        }
        let parsed = parse_row_cells(&ln.text, ln.start_offset, source);
        rows.push(Row {
            cells: parsed.cells,
            cell_spans: parsed.cell_spans,
            span: Span::from_offsets(source, ln.start_offset, ln.end_offset),
        });
        i += 1;
    }
    let end_offset = rows
        .last()
        .map_or(delim_line.end_offset, |r| r.span.end_offset);
    let table = Table {
        span: Span::from_offsets(source, header_line.start_offset, end_offset),
        header,
        rows,
    };
    Some((table, i))
}

File: go/core/scanner.go
package core

import (
	"regexp"
	"strings"
)

// Turns raw Markdown into a flat list of Block nodes — port of scanner.ts /
// scanner.rs. Offsets in stored spans are UTF-16 code units; the line splitter
// keeps a running (byte, UTF-16) dual cursor and per-line regex offsets are
// converted from bytes to UTF-16.

// rawLine is one line of source, with its UTF-16 and byte offsets in the source.
type rawLine struct {
	text        string
	startOffset int
	endOffset   int
	startByte   int
	endByte     int
}

// The `\1` backreference is expanded into three alternatives (RE2 has no
// backreferences); otherwise these mirror the reference patterns.
var (
	thematicRE    = regexp.MustCompile(`^\s*(?:-(?:\s*-){2,}|\*(?:\s*\*){2,}|_(?:\s*_){2,})\s*$`)
	ulRE          = regexp.MustCompile(`^(\s*)([-*+])\s+(.*)$`)
	olRE          = regexp.MustCompile(`^(\s*)([0-9]+)([.)])\s+(.*)$`)
	bqRE          = regexp.MustCompile(`^>\s?(.*)$`)
	fenceRE       = regexp.MustCompile("^(`{3,})\\s*(\\S*)\\s*$")
	rowRE         = regexp.MustCompile(`^\|(.+)\|\s*$`)
	delimRE       = regexp.MustCompile(`^\|\s*:?-+:?\s*(\|\s*:?-+:?\s*)*\|\s*$`)
	headingRE     = regexp.MustCompile(`^(#{1,6})\s+(.*?)(?:\s+#+)?\s*$`)
	headingPrefRE = regexp.MustCompile(`^#{1,6}\s+`)
)

// scan scans source into a list of Block nodes.
func scan(source string) []Block {
	lines := splitLines(source)
	var blocks []Block

	i := 0
	for i < len(lines) {
		if javaTrim(lines[i].text) == "" {
			i++
			continue
		}
		if fence, next, ok := tryFence(source, lines, i); ok {
			blocks = append(blocks, fence)
			i = next
			continue
		}
		if table, next, ok := tryTable(source, lines, i); ok {
			blocks = append(blocks, table)
			i = next
			continue
		}
		if tb, ok := tryThematicBreak(source, lines[i]); ok {
			blocks = append(blocks, tb)
			i++
			continue
		}
		if quote, next, ok := tryBlockquote(source, lines, i); ok {
			blocks = append(blocks, quote)
			i = next
			continue
		}
		if heading, ok := tryHeading(source, lines[i]); ok {
			blocks = append(blocks, heading)
			i++
			continue
		}
		if item, ok := tryListItem(source, lines[i]); ok {
			blocks = append(blocks, item)
			i++
			continue
		}
		paragraph, next := consumeParagraph(source, lines, i)
		blocks = append(blocks, paragraph)
		i = next
	}
	return blocks
}

func splitLines(source string) []rawLine {
	var out []rawLine
	byteStart := 0
	u16Start := 0
	u16 := 0
	for byteI, c := range source {
		if c == '\n' {
			out = append(out, rawLine{
				text:        source[byteStart:byteI],
				startOffset: u16Start,
				endOffset:   u16,
				startByte:   byteStart,
				endByte:     byteI,
			})
			byteStart = byteI + 1
			u16Start = u16 + 1
		}
		u16 += utf16RuneLen(c)
	}
	out = append(out, rawLine{
		text:        source[byteStart:],
		startOffset: u16Start,
		endOffset:   u16,
		startByte:   byteStart,
		endByte:     len(source),
	})
	return out
}

func tryThematicBreak(source string, line rawLine) (Block, bool) {
	if !thematicRE.MatchString(line.text) {
		return nil, false
	}
	return ThematicBreak{Span: spanFromOffsets(source, line.startOffset, line.endOffset)}, true
}

func tryHeading(source string, line rawLine) (Block, bool) {
	m := headingRE.FindStringSubmatch(line.text)
	if m == nil {
		return nil, false
	}
	hashes := m[1]
	text := javaTrim(m[2])
	return Heading{
		Level: len(hashes),
		Text:  text,
		Span:  spanFromOffsets(source, line.startOffset, line.endOffset),
	}, true
}

func tryListItem(source string, line rawLine) (Block, bool) {
	if ul := ulRE.FindStringSubmatch(line.text); ul != nil {
		text := ul[3]
		markerStart := line.startOffset + utf16Len(ul[1])
		markerEnd := markerStart + utf16Len(ul[2])
		textStart := line.startOffset + utf16Index(line.text, strings.Index(line.text, text))
		return ListItem{
			Text:       text,
			Span:       spanFromOffsets(source, line.startOffset, line.endOffset),
			SegmentMap: []SegmentOffset{{TextOffset: 0, SourceOffset: textStart}},
			Ordered:    false,
			MarkerSpan: spanFromOffsets(source, markerStart, markerEnd),
		}, true
	}
	if ol := olRE.FindStringSubmatch(line.text); ol != nil {
		text := ol[4]
		markerStart := line.startOffset + utf16Len(ol[1])
		markerEnd := markerStart + utf16Len(ol[2]) + utf16Len(ol[3])
		textStart := line.startOffset + utf16Index(line.text, strings.Index(line.text, text))
		return ListItem{
			Text:       text,
			Span:       spanFromOffsets(source, line.startOffset, line.endOffset),
			SegmentMap: []SegmentOffset{{TextOffset: 0, SourceOffset: textStart}},
			Ordered:    true,
			MarkerSpan: spanFromOffsets(source, markerStart, markerEnd),
		}, true
	}
	return nil, false
}

func tryBlockquote(source string, lines []rawLine, startIdx int) (Block, int, bool) {
	first := lines[startIdx]
	m := bqRE.FindStringSubmatch(first.text)
	if m == nil {
		return nil, 0, false
	}
	firstSegment := m[1]

	segments := []string{firstSegment}
	segmentMap := []SegmentOffset{{
		TextOffset:   0,
		SourceOffset: first.startOffset + utf16Index(first.text, strings.Index(first.text, firstSegment)),
	}}
	joinedTextOffset := utf16Len(firstSegment)

	i := startIdx + 1
	endOffset := first.endOffset
	for i < len(lines) {
		ln := lines[i]
		next := bqRE.FindStringSubmatch(ln.text)
		if next == nil {
			break
		}
		segment := next[1]
		joinedTextOffset++ // newline separator
		segmentMap = append(segmentMap, SegmentOffset{
			TextOffset:   joinedTextOffset,
			SourceOffset: ln.startOffset + utf16Index(ln.text, strings.Index(ln.text, segment)),
		})
		joinedTextOffset += utf16Len(segment)
		segments = append(segments, segment)
		endOffset = ln.endOffset
		i++
	}
	quote := Blockquote{
		Text:       strings.Join(segments, "\n"),
		Span:       spanFromOffsets(source, first.startOffset, endOffset),
		SegmentMap: segmentMap,
	}
	return quote, i, true
}

func consumeParagraph(source string, lines []rawLine, startIdx int) (Block, int) {
	first := lines[startIdx]
	endIdx := startIdx
	for endIdx+1 < len(lines) {
		t := lines[endIdx+1].text
		if javaTrim(t) == "" ||
			headingPrefRE.MatchString(t) ||
			ulRE.MatchString(t) ||
			olRE.MatchString(t) ||
			bqRE.MatchString(t) ||
			fenceRE.MatchString(t) ||
			rowRE.MatchString(t) ||
			thematicRE.MatchString(t) {
			break
		}
		endIdx++
	}
	last := lines[endIdx]
	paragraph := Paragraph{
		Text:       source[first.startByte:last.endByte],
		Span:       spanFromOffsets(source, first.startOffset, last.endOffset),
		SegmentMap: []SegmentOffset{{TextOffset: 0, SourceOffset: first.startOffset}},
	}
	return paragraph, endIdx + 1
}

func tryFence(source string, lines []rawLine, startIdx int) (Block, int, bool) {
	start := lines[startIdx]
	open := fenceRE.FindStringSubmatch(start.text)
	if open == nil {
		return nil, 0, false
	}
	fenceMarker := open[1]
	info := javaTrim(open[2])

	i := startIdx + 1
	var bodyStartU16, bodyStartByte int
	haveBodyStart := false
	var bodyEndU16, bodyEndByte int
	haveBodyEnd := false
	endOffset := start.endOffset
	for i < len(lines) {
		ln := lines[i]
		if close := fenceRE.FindStringSubmatch(ln.text); close != nil {
			if len(close[1]) >= len(fenceMarker) {
				endOffset = ln.endOffset
				break
			}
		}
		if !haveBodyStart {
			bodyStartU16 = ln.startOffset
			bodyStartByte = ln.startByte
			haveBodyStart = true
		}
		// Include the newline that separates this line from the next.
		bodyEndU16 = ln.endOffset + 1
		bodyEndByte = ln.endByte + 1
		haveBodyEnd = true
		i++
	}

	sourceU16 := utf16Len(source)
	clampedEndU16 := 0
	clampedEndByte := 0
	if haveBodyEnd {
		clampedEndU16 = min(bodyEndU16, sourceU16)
		clampedEndByte = min(bodyEndByte, len(source))
	}
	body := ""
	if haveBodyStart && haveBodyEnd {
		body = source[bodyStartByte:clampedEndByte]
	}
	fallback := start.endOffset
	bodyStartSpanU16 := fallback
	if haveBodyStart {
		bodyStartSpanU16 = bodyStartU16
	}
	bodyEndSpanU16 := fallback
	if haveBodyEnd {
		bodyEndSpanU16 = clampedEndU16
	}
	fence := Fence{
		Span:     spanFromOffsets(source, start.startOffset, endOffset),
		Info:     info,
		Body:     body,
		BodySpan: spanFromOffsets(source, bodyStartSpanU16, bodyEndSpanU16),
	}
	return fence, i + 1, true
}

func tryTable(source string, lines []rawLine, startIdx int) (Block, int, bool) {
	if startIdx+1 >= len(lines) {
		return nil, 0, false
	}
	headerLine := lines[startIdx]
	delimLine := lines[startIdx+1]
	if !rowRE.MatchString(headerLine.text) || !delimRE.MatchString(delimLine.text) {
		return nil, 0, false
	}

	headerParsed := parseRowCells(headerLine.text, headerLine.startOffset, source)
	header := Row{
		Cells:     headerParsed.Cells,
		CellSpans: headerParsed.CellSpans,
		Span:      spanFromOffsets(source, headerLine.startOffset, headerLine.endOffset),
	}

	var rows []Row
	i := startIdx + 2
	for i < len(lines) {
		ln := lines[i]
		if !rowRE.MatchString(ln.text) {
			break
		}
		parsed := parseRowCells(ln.text, ln.startOffset, source)
		rows = append(rows, Row{
			Cells:     parsed.Cells,
			CellSpans: parsed.CellSpans,
			Span:      spanFromOffsets(source, ln.startOffset, ln.endOffset),
		})
		i++
	}
	endOffset := delimLine.endOffset
	if len(rows) > 0 {
		endOffset = rows[len(rows)-1].Span.EndOffset
	}
	table := Table{
		Span:   spanFromOffsets(source, headerLine.startOffset, endOffset),
		Header: header,
		Rows:   rows,
	}
	return table, i, true
}
